File: CPP Codes/12_H_Index.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        vector<int> cnt(n + 1, 0);
        for (int x : citations) {
            ++cnt[min(x, n)];
        }
        int s = 0;
        for (int h = n; h >= 0; --h) {
            s += cnt[h];
            if (s >= h) {
                return h;
            }
        }
        return 0;  // In case no valid h-index is found
    }
};

int main() {
    vector<int> citations = {3, 0, 6, 1, 5};
    
    Solution solution;
    int hIndex = solution.hIndex(citations);
    cout << "The H-Index is: " << hIndex << endl;

    return 0;
}

File: Python Codes/12_H_Index.py
class Solution:
    def hIndex(self, citations):
        n = len(citations)
        cnt = [0] * (n + 1)
        for x in citations:
            cnt[min(x, n)] += 1
        s = 0
        for h in range(n, -1, -1):
            s += cnt[h]
            if s >= h:
                return h
        return 0  # In case no valid h-index is found

if __name__ == "__main__":
    citations = [3, 0, 6, 1, 5]
    
    solution = Solution()
    hIndex = solution.hIndex(citations)
    print(f"The H-Index is: {hIndex}")
